cotix/_collision_resolution.py

```diff
@@ -13,68 +13,253 @@
 
-    v1_new = jnp.matmul(change_of_basis_inv, v1qr_new)
-    v2_new = jnp.matmul(change_of_basis_inv, v2qr_new)
+    # called iff velocities_away is False
+    def continue_resolution(body1, body2):
+        # impulse computation is in accordance with
+        # https://github.com/knyazer/RSS/blob/
+        # 1246e03c5950a5549a128fbce97c7bd402f9bed7/engine/source/env/World.cpp#L87
 
-    body1 = body1.set_velocity(v1_new)
-    body2 = body2.set_velocity(v2_new)
+        # inertia is kg * m^2, so this is kg^-1
+        impulseFactor1 = (1 / body1.mass) + (lever_arm1**2) / body1.inertia
+        impulseFactor2 = (1 / body2.mass) + (lever_arm2**2) / body2.inertia
 
-    body1, body2 = _split_bodies(body1, body2, epa_vector)
-    return body1, body2
+        # here, we only care about the collision along the collision normal
+        # units are kg * m / s
+        primary_col_impulse = (
+            (1 + elasticity) * v_rel[0] / (impulseFactor1 + impulseFactor2)
+        )
+
+        # v1 dot perpendicular tries to be the same as v2 dot perpendicular.
+        #   clipped to be in the range [-friction_impulse_max, friction_impulse_max]
+        friction_impulse_max = friction_coefficient * jnp.abs(primary_col_impulse)
+        # this adds an impulse perpendicular to the collision normal
+        friction_impulse = jnp.clip(
+            # I am not too sure that the formula is correct here
+            tangential_relative_velocity / (impulseFactor1 + impulseFactor2),
+            -friction_impulse_max,
+            friction_impulse_max,
+        )
+
+        col_impulse = -jnp.array([primary_col_impulse, friction_impulse])
 
+        # jax.debug.print(
+        #     "\nlever arms: {lever_arm1}, {lever_arm2}. \n"
+        #     "center1: {center1}, center2: {center2}. "
+        #     "contact_point: {contact_point}. \n"
+        #     "global_supports {sup1}, {sup2}. "
+        #     "collision_unit_vector {unit_collision_vector}. \n"
+        #     "tangential_relative_velocity: {tangential_relative_velocity}. \n"
+        #     "friction_impulse: {friction_impulse}. \n"
+        #     "col_impulse in new basis: {col_impulse}. \n"
+        #     "col_impulse original: {col_impulse_original}. \n",
+        #     lever_arm1=lever_arm1,
+        #     lever_arm2=lever_arm2,
+        #     center1=body1.get_center_of_mass(),
+        #     center2=body2.get_center_of_mass(),
+        #     contact_point=contact_point,
+        #     sup1=body1.shape.get_global_support(-unit_collision_vector),
+        #     sup2=body2.shape.get_global_support(unit_collision_vector),
+        #     unit_collision_vector=unit_collision_vector,
+        #     tangential_relative_velocity=tangential_relative_velocity,
+        #     friction_impulse=friction_impulse,
+        #     col_impulse=col_impulse,
+        #     col_impulse_original=change_of_basis_inv @ col_impulse,
+        # )
```

Comment:
you can use jax.debug.breakpoint or eqx.debug.breakpoint_if, it is easier somewhat